tablo/geom_utils.py

```diff
@@ -437,4 +437,196 @@ def extract_significant_digits(number):
+
+    @staticmethod
+    def rings_to_wkt(rings):
+        outer_rings = []
+        holes = []
+
+        for r in rings:
+            ring = Terraformer.close_ring(r)
+            if len(ring) < 4:
+                continue
+
+            if Terraformer.ring_is_clockwise(ring):
+                # a clockwise ring means its an outer ring
+                polygon = [ring]
+                outer_rings.append(polygon)
+            else:
+                # otherwise it's a hole!
+                holes.append(ring)
+
+        uncontained_holes = []
+
+        while holes:
+            hole = holes.pop()
+
+            # loop over all outer rings and see if they contain our hole.
+            contained = False
+            for r in reversed(outer_rings):     # TODO why esri loops over a reversed rings?
+                outer_ring = r[0]
+                if Terraformer.coordinates_contain_coordinates(outer_ring, hole):
+                    r.append(hole)
+                    contained = True
+                    break
+
+            # sometimes a ring may not be contained in any outer ring (https://github.com/Esri/esri-leaflet/issues/320)
+            if not contained:
+                uncontained_holes.append(hole)
+
+        # check and see if uncontained holes intersect with any outer rings
+        while uncontained_holes:
+            hole = uncontained_holes.pop()
+
+            intersects = False
+            for r in reversed(outer_rings):     # TODO why esri loops over a reversed rings?
+                outer_ring = r[0]
+                if Terraformer.arrays_intersect_arrays(outer_ring, hole):
+                    outer_rings[r].push(hole)
+                    intersects = True
+                    break
+
+            # hole does not intersect ANY outer ring at this point; make it an outer ring.
+            if not intersects:
+                hole.reverse()
+                outer_rings.append([hole])
+
+        return {
+            'type': 'MultiPolygon',
+            'coordinates': outer_rings
+        }
+
+    @staticmethod
+    def close_ring(coordinates):
+        # makes sure the ring is closed
+        if coordinates[0] != coordinates[-1]:
+            coordinates.push(coordinates[0])
+
+        return coordinates
+
+    @staticmethod
+    def ring_is_clockwise(ring_to_test):
+        # a positive sum of areas under each side of a polygon indicates a clockwise polygon
+        # http://stackoverflow.com/questions/1165647/how-to-determine-if-a-list-of-polygon-points-are-in-clockwise-order
+        total = 0
+        r_length = len(ring_to_test)
+        for idx, pt1 in enumerate(ring_to_test):
+            if idx + 1 != r_length:
+                pt2 = ring_to_test[idx + 1]
+                total += (pt2[0] - pt1[0]) * (pt2[1] + pt1[1])
+        return total >= 0
+
+    @staticmethod
+    def edge_intersects_edge(a1, a2, b1, b2):
+        ua_t = (b2[0] - b1[0]) * (a1[1] - b1[1]) - (b2[1] - b1[1]) * (a1[0] - b1[0])
```

Comment:
It may be a good idea to enforce precedence with parens, just for clarity sake. For example:

```python
        ua_t = ((b2[0] - b1[0]) * (a1[1] - b1[1])) - ((b2[1] - b1[1]) * (a1[0] - b1[0]))
```

Also, if there is a way to aptly name the components of these equations, it might be a good idea to break them out as sub-variables (not sure what the best names would be):

```python
        edge_start_diff = (a1[0] - b1[0])
        edge_end_diff = (a1[1] - b1[1])
        edge_two_start_diff = (b2[0] - b1[0])
        edge_two_end_diff = (b2[1] - b1[1])
        ...
```

...anyway, just for consideration. It looks very efficient I'm just not sure when I look at it what is happening.